File: packages/python/cerebrum_sdk/chain.py
"""Chain builder for composing blocks."""

from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .client import Cerebrum


class Chain:
    """Builder for chaining multiple blocks together.
    
    Chain blocks to create powerful AI pipelines:
    PDF → OCR → Chat = Document AI in 3 lines
    
    Example:
        >>> from cerebrum_sdk import Cerebrum, chain
        >>> 
        >>> client = Cerebrum(api_key="cb_your_key")
        >>> 
        >>> result = await chain(client) \
        ...     .then("pdf", {"extract": "text"}) \
        ...     .then("translate", {"target": "en"}) \
        ...     .then("chat", {"prompt": "Summarize:"}) \
        ...     .run("document.pdf")
        >>> 
        >>> print(result.final_output)
    """
    
    def __init__(self, client: "Cerebrum"):
        self.client = client
        self.steps: List[Dict[str, Any]] = []
    
    def then(self, block: str, params: Optional[Dict[str, Any]] = None) -> "Chain":
        """Add a block to the chain.
        
        Args:
            block: Block name (chat, pdf, ocr, translate, etc.)
            params: Parameters for this block
        
        Returns:
            Self for method chaining
        
        Example:
            >>> chain.then("pdf", {"extract": "text"})
            >>> chain.then("chat", {"prompt": "Summarize:"})
        """
        self.steps.append({
            "block": block,
            "params": params or {}
        })
        return self
    
    async def run(self, initial_input: Any) -> "ChainResult":
        """Execute the chain with initial input.
        
        Args:
            initial_input: Starting input (file path, text, etc.)
        
        Returns:
            ChainResult with final_output and step_results
        """
        result = await self.client.chain(self.steps, initial_input)
        return ChainResult(result)


class ChainResult:
    """Result of a chain execution.
    
    Attributes:
        success: Whether all steps completed successfully
        final_output: The final output from the chain
        steps_executed: Number of steps that ran
        total_time_ms: Total execution time in milliseconds
        step_results: Detailed results from each step
    """
    
    def __init__(self, data: Dict[str, Any]):
        self._data = data
    
    @property
    def success(self) -> bool:
        """Whether all steps completed successfully."""
        return self._data.get("status") == "success"
    
    @property
    def final_output(self) -> Any:
        """The final output from the chain."""
        return self._data.get("final_output")
    
    @property
    def steps_executed(self) -> int:
        """Number of steps that ran."""
        return self._data.get("steps_executed", 0)
    
    @property
    def total_time_ms(self) -> int:
        """Total execution time in milliseconds."""
        return self._data.get("total_time_ms", 0)
    
    @property
    def step_results(self) -> List[Dict[str, Any]]:
        """Detailed results from each step."""
        return self._data.get("step_results", [])
    
    @property
    def raw(self) -> Dict[str, Any]:
        """Raw response data."""
        return self._data
    
    def __repr__(self) -> str:
        return f"ChainResult(success={self.success}, steps={self.steps_executed})"


def chain(client: "Cerebrum") -> Chain:
    """Create a new chain builder.
    
    Args:
        client: Cerebrum client instance
    
    Returns:
        Chain builder for composing blocks
    
    Example:
        >>> from cerebrum_sdk import Cerebrum, chain
        >>> client = Cerebrum(api_key="cb_your_key")
        >>> 
        >>> result = await chain(client) \
        ...     .then("pdf", {"extract": "text"}) \
        ...     .then("chat", {"prompt": "Summarize:"}) \
        ...     .run("document.pdf")
    """
    return Chain(client)

File: packages/js/src/chain.ts
/**
 * Chain builder for composing blocks
 */

import { Cerebrum } from './client';
import { ChainStep } from './types';

/** Result of a chain execution */
export class ChainResult {
  private _data: Record<string, any>;
  
  constructor(data: Record<string, any>) {
    this._data = data;
  }
  
  /** Whether all steps completed successfully */
  get success(): boolean {
    return this._data.status === 'success';
  }
  
  /** The final output from the chain */
  get finalOutput(): any {
    return this._data.final_output;
  }
  
  /** Number of steps that ran */
  get stepsExecuted(): number {
    return this._data.steps_executed || 0;
  }
  
  /** Total execution time in milliseconds */
  get totalTimeMs(): number {
    return this._data.total_time_ms || 0;
  }
  
  /** Detailed results from each step */
  get stepResults(): any[] {
    return this._data.step_results || [];
  }
  
  /** Raw response data */
  get raw(): Record<string, any> {
    return this._data;
  }
  
  toJSON(): Record<string, any> {
    return this._data;
  }
}

/** Builder for chaining multiple blocks together */
export class Chain {
  private client: Cerebrum;
  private steps: ChainStep[] = [];
  
  constructor(client: Cerebrum) {
    this.client = client;
  }
  
  /**
   * Add a block to the chain
   * @param block - Block name (chat, pdf, ocr, translate, etc.)
   * @param params - Parameters for this block
   * @returns Self for method chaining
   * 
   * @example
   * chain.then("pdf", { extract: "text" })
   * chain.then("chat", { prompt: "Summarize:" })
   */
  then(block: string, params?: Record<string, any>): Chain {
    this.steps.push({ block, params });
    return this;
  }
  
  /**
   * Execute the chain with initial input
   * @param initialInput - Starting input (file path, text, etc.)
   * @returns ChainResult with final_output and step_results
   */
  async run(initialInput: any): Promise<ChainResult> {
    const result = await this.client.chain(this.steps, initialInput);
    return new ChainResult(result);
  }
}

/**
 * Create a new chain builder
 * @param client - Cerebrum client instance
 * @returns Chain builder for composing blocks
 * 
 * @example
 * const result = await chain(client)
 *   .then("pdf", { extract: "text" })
 *   .then("chat", { prompt: "Summarize:" })
 *   .run("document.pdf");
 */
export function chain(client: Cerebrum): Chain {
  return new Chain(client);
}
